### PR TITLE
Move the `assets` folder to `public/assets`

### LINKED ISSUE #7093: Stop using the `assets` folder
### Description

Our goal is not to use the `assets` folder anymore, because it is "reserved" for the Symfony app. Therefore:

1. Install the Contao components into `public/components`.

2. Move the `assets/{css,images,js}` folders to `public/assets/{css,images,js}`, so the existing paths in CDNs and other caches don‘t change.

Also see #7013.

### PR DESCRIPTION
**Pull request by leofeyer**
Implements #7093

With these surprisingly few changes, we can get rid of the `assets` directory in the project root, as the assets and Contao components are installed directly in the `public/assets` folder.

I know that we originally wanted to install the Contao components in `public/components`, but that would break all existing references in `$GLOBALS['TL_CSS']`, `$GLOBALS['TL_JAVASCRIPT']` and so on. The current solution is almost 100% backwards compatible because no paths change at all.

### CODE REVIEW

File: core-bundle/contao/library/Contao/Combiner.php [hunk 1]
```diff
@@ -210,10 +193,10 @@ public function getFileUrls($strUrl=null)
...
 				$strPath = 'assets/' . $strTarget . '/' . str_replace('/', '_', $arrFile['name']) . $this->strMode;
 
-				if ($blnDebug || !file_exists($this->strRootDir . '/' . $strPath))
+				if ($blnDebug || !file_exists($this->strWebDir . '/' . $strPath))
 				{
-					$objFile = new File($strPath);
-					$objFile->write($this->handleScssLess(file_get_contents($this->strRootDir . '/' . $arrFile['name']), $arrFile));
+					$objFile = new File(StringUtil::stripRootDir($this->strWebDir . '/' . $strPath));
+					$objFile->write($this->handleScssLess(file_get_contents($this->strWebDir . '/' . $arrFile['name']), $arrFile));
 					$objFile->close();
```
Comment (leofeyer) [inline core-bundle/contao/library/Contao/Combiner.php]: We should probably replace this with the Symfony filesystem.

File: core-bundle/contao/library/Contao/Combiner.php [hunk 2]
```diff
@@ -319,18 +296,18 @@ protected function getCombinedFileUrl($strUrl=null)
...
-		if (file_exists($this->strRootDir . '/assets/' . $strTarget . '/' . $strKey . $this->strMode))
+		if (file_exists($this->strWebDir . '/assets/' . $strTarget . '/' . $strKey . $this->strMode))
 		{
 			return $strUrl . 'assets/' . $strTarget . '/' . $strKey . $this->strMode;
 		}
 
 		// Create the file
-		$objFile = new File('assets/' . $strTarget . '/' . $strKey . $this->strMode);
+		$objFile = new File(StringUtil::stripRootDir($this->strWebDir . '/assets/' . $strTarget . '/' . $strKey . $this->strMode));
 		$objFile->truncate();
```
Comment (leofeyer) [inline core-bundle/contao/library/Contao/Combiner.php]: Same here (Symfony filesystem).

File: core-bundle/contao/library/Contao/Image.php [hunk 3]
```diff
@@ -93,7 +93,7 @@ public static function getPath($src)
 			$filename = pathinfo($src, PATHINFO_FILENAME);
 
 			// Prefer SVG icons
-			if (file_exists($projectDir . '/assets/contao/images/' . $filename . '.svg'))
+			if (file_exists($webDir . '/assets/contao/images/' . $filename . '.svg'))
```
Comment (leofeyer) [inline core-bundle/contao/library/Contao/Image.php]: This is actually a Contao 2 BC layer. Removing it would save us a lot of `file_exists()` calls.

Comment (leofeyer) [reply core-bundle/contao/library/Contao/Image.php]: Removed in 60f9b4a24d39a7ea3ee4fc2c203f30c6bfa3d858.

File: core-bundle/contao/library/Contao/Combiner.php (hunk 1 above)
Comment (leofeyer) [reply core-bundle/contao/library/Contao/Combiner.php]: Replaced in 3d53ece8c5598ba4e44f7fe477168469ffdf8df2.

File: core-bundle/contao/library/Contao/Combiner.php (hunk 2 above)
Comment (leofeyer) [reply core-bundle/contao/library/Contao/Combiner.php]: See 3d53ece8c5598ba4e44f7fe477168469ffdf8df2

**Comment by aschempp** [comment]
How does this handle the case that any existing Contao installation has `contao-component-dir` set to `assets` and not `public/assets`?

**Comment by leofeyer** [comment]
There are two possible solutions:

1. We update the `composer.json` file in the manager (preferred).
2. We keep symlinking the `assets` directory if `contao-component-dir` is set to `assets`.

File: core-bundle/contao/classes/PurgeData.php [hunk 4]
```diff
@@ -102,8 +102,12 @@ public function run()
 			// Get the current folder size
 			foreach ($config['affected'] as $folder)
 			{
+				if (str_contains($folder, '%'))
+				{
+					$folder = StringUtil::stripRootDir($parameterBag->resolveValue($folder));
+				}
+
 				$total = 0;
-				$folder = sprintf($folder, $strCachePath);
```
Comment (ausi) [inline core-bundle/contao/classes/PurgeData.php:106]: Isn’t this a BC break if someone used this previously like 
`$GLOBALS['TL_PURGE']['folders']['foo']['affected'] = array('%s/foo/bar')`?
Or was this never officially supported?

File: core-bundle/contao/library/Contao/Combiner.php [hunk 5]
```diff
@@ -122,17 +120,9 @@ public function add($strFile, $strVersion=null, $strMedia='all')
...
-			if (file_exists($this->strRootDir . '/' . $this->strWebDir . '/' . $strFile))
-			{
-				$strFile = $this->strWebDir . '/' . $strFile;
-			}
-			else
-			{
-				return;
-			}
+			return;
 		}
```
Comment (ausi) [inline core-bundle/contao/library/Contao/Combiner.php]: This would break BC for adding files that are not symlinked to the web dir I think.
For example SCSS files from a non-public `files/` folder. _(Did not test that manually yet)_

File: core-bundle/contao/library/Contao/Image.php [hunk 6]
```diff
@@ -81,23 +81,8 @@ public static function getPath($src)
...
-				return 'assets/contao/images/' . $src;
-			}
-
-			$filename = pathinfo($src, PATHINFO_FILENAME);
-
-			// Prefer SVG icons
-			if (file_exists($projectDir . '/assets/contao/images/' . $filename . '.svg'))
-			{
-				return 'assets/contao/images/' . $filename . '.svg';
-			}
```
Comment (ausi) [inline core-bundle/contao/library/Contao/Image.php:106]: Seems to be unrelated to this PR and breaks BC.

File: core-bundle/contao/library/Contao/Image.php [hunk 7]
```diff
@@ -113,13 +98,7 @@ public static function getPath($src)
 			trigger_deprecation('contao/core-bundle', '5.2', 'Using the "%s" icon has been deprecated and will no longer work in Contao 6. Use the "%s--disabled" icon instead.', $filename, substr($filename, 0, -1));
 		}
 
-		// Prefer SVG icons
-		if (file_exists($projectDir . '/system/themes/' . $theme . '/icons/' . $filename . '.svg'))
-		{
-			return 'system/themes/' . $theme . '/icons/' . $filename . '.svg';
-		}
```
Comment (ausi) [inline core-bundle/contao/library/Contao/Image.php]: Seems to be unrelated to this PR and breaks BC.

File: core-bundle/tests/Contao/CombinerTest.php [hunk 8]
```diff
@@ -143,9 +140,9 @@ public function testHandlesSpecialCharactersWhileFixingTheFilePaths(): void
             EOF;
 
         $expected = <<<'EOF'
-            test1 { background: url("../../\"test\"/foo.bar") }
-            test2 { background: url("../../\"test\"/foo.bar") }
-            test3 { background: url('../../"test"/foo.bar') }
+            test1 { background: url("../../public/\"test\"/foo.bar") }
+            test2 { background: url("../../public/\"test\"/foo.bar") }
+            test3 { background: url('../../public/"test"/foo.bar') }
```
Comment (ausi) [inline core-bundle/tests/Contao/CombinerTest.php]: I’d change the `['name' => 'public/"test"/file.css']` to `['name' => '"test"/file.css']` a few lines below and revert these changes here to make the intended behavior more clear.
But this is just a nitpick, feel free to ignore.

**Review by ausi — COMMENTED** [review]
I think we should improve BC a bit, otherwise this is the way to go IMO 👍

File: core-bundle/contao/classes/PurgeData.php (hunk 4 above)
Comment (leofeyer) [reply core-bundle/contao/classes/PurgeData.php:106]: That was never officially supported.

File: core-bundle/contao/library/Contao/Image.php (hunk 6 above)
Comment (leofeyer) [reply core-bundle/contao/library/Contao/Image.php:106]: As I wrote, this was a Contao 2 BC layer that we forgot to remove. The break would only affect extensions that are compatible with Contao 2 to 5, so effectively none.

File: core-bundle/contao/library/Contao/Image.php (hunk 7 above)
Comment (leofeyer) [reply core-bundle/contao/library/Contao/Image.php]: See https://github.com/contao/contao/pull/7094#discussion_r1611746056

File: core-bundle/contao/library/Contao/Image.php (hunk 6 above)
Comment (ausi) [reply core-bundle/contao/library/Contao/Image.php:106]: My extensions still use `.gif` for icons that did not change since Contao 3 🙈
I should fix that anyway but I think the possibility of other extensions having the same issue is quite high.

Comment (leofeyer) [reply core-bundle/contao/library/Contao/Image.php:106]: Okay, I guess in this case we have to keep the BC layer. I will adjust the PR accordingly.

File: core-bundle/contao/library/Contao/Combiner.php (hunk 5 above)
Comment (leofeyer) [reply core-bundle/contao/library/Contao/Combiner.php]: Reverted in 5e074678b80dacda008271d092e1725d3b6c3951.

File: core-bundle/tests/Contao/CombinerTest.php (hunk 8 above)
Comment (leofeyer) [reply core-bundle/tests/Contao/CombinerTest.php]: Changed in 5e074678b80dacda008271d092e1725d3b6c3951.

File: core-bundle/contao/library/Contao/Image.php (hunk 6 above)
Comment (leofeyer) [reply core-bundle/contao/library/Contao/Image.php:106]: Now that I‘ve taken a closer look, I don‘t think we need to keep this.

* Let‘s assume your code does `Image::getPath('icon.gif')`.
* Let‘s also assume that both `icon.gif` and `icon.svg` exist.
* In this special case, the BC layer would return `assets/contao/images/icon.svg`, even if you requested `icon.gif`.
* Without the BC layer, the method would simply return `assets/contao/images/icon.gif` as requested.

So in both cases an icon is returned. Do we really want to keep the additional `file_exists()` call just to return an SVG instead of a GIF when both icons exist?

**Comment by leofeyer** [comment]
> How does this handle the case that any existing Contao installation has `contao-component-dir` set to `assets` and not `public/assets`?

@aschempp I have added a BC layer in e41594b0b3e6f887e2dc48019a4570c983d79703. I would still prefer the manager to update the `composer.json` file, if that‘s possible.

File: core-bundle/src/Command/SymlinksCommand.php [hunk 9]
```diff
@@ -87,8 +87,13 @@ private function generateSymlinks(): void
         $this->symlinkModules();
         $this->symlinkThemes();
 
-        // Symlink the assets and themes directory
-        $this->symlink('assets', Path::join($this->webDir, 'assets'));
+        // Symlink the assets directory if the Contao components were installed there
+        // instead of in public/assets (backwards compatibility)
+        if (is_dir($this->projectDir.'/assets/contao')) {
```
Comment (leofeyer) [inline core-bundle/src/Command/SymlinksCommand.php]: This will break if someone creates a `contao` subfolder in their `assets` directory.  We could check for multiple components instead, e.g. `assets/dropzone` and `assets/handorgel`, but that would also break if we remove one of them in the future.

@ausi Can you think of a better check that does not require reading the `composer.json` configuration?

File: core-bundle/contao/library/Contao/Image.php (hunk 6 above)
Comment (ausi) [reply core-bundle/contao/library/Contao/Image.php:106]: > Let‘s also assume that both `icon.gif` and `icon.svg` exist.

In the core all `.gif` icons were removed, so things like `edit.gif` would not work anymore.

> So in both cases an icon is returned. Do we really want to keep the additional `file_exists()` call just to return an SVG instead of a GIF when both icons exist?

IMO yes. The extra `file_exists()` check is only needed if the requested file is not an `.svg` already, so it would only affect the _wrong_ or _old_ calls.

File: core-bundle/src/Command/SymlinksCommand.php (hunk 9 above)
Comment (ausi) [reply core-bundle/src/Command/SymlinksCommand.php]: Check the other way round?
```suggestion
        if (!is_dir(Path::join($this->webDir, 'assets')) || is_link(Path::join($this->webDir, 'assets'))) {
```
Because if the components get installed to `public/assets` the directory would already be there at this point I guess?

But as this command is not performance relevant, I’d prefer reading the composer.json instead.

File: core-bundle/contao/library/Contao/Image.php (hunk 6 above)
Comment (leofeyer) [reply core-bundle/contao/library/Contao/Image.php:106]: Did you realize that the code only handles icons from `contao-components/contao` that begin with `icon`? Things like `edit.gif` are not processed here. And since there are no .gif icons in the component anymore, I think we can safely remove this BC layer (but not the one below).

File: core-bundle/contao/library/Contao/Image.php (hunk 7 above)
Comment (leofeyer) [reply core-bundle/contao/library/Contao/Image.php]: I agree that we have to keep this BC layer for old theme icons.

**Comment by richardhj** [comment]
In #7093 we talked about that the `assets` and `public/assets` folders are reserved by the Symfony framework.
I guess the solution, for those using a Symfony app, is to use a different folder for Symfony's version of assets..?

Is becoming compatible with symfony/asset-mapper not an alternative goal here?

**Comment by leofeyer** [comment]
It is, and that‘s exactly what the pull request does. After it has been merged, Contao will no longer use the `assets` folder. But of course, `public/assets` must be shared by Contao and Symfony to avoid BC breaks.

```
assets/                <-- available for the Symfony asset mapper

public/assets/
    ace/               <-- Contao component
    css/               <-- Contao
    images/            <-- Contao
    js/                <-- Contao
    styles/            <-- Symfony
    app.js             <-- Symfony
```

Anything beyond that can only be done in Contao 6.

**Comment by richardhj** [comment]
If there are no naming conflicts, that would work fine! Otherwise one could always also change the output folder of asset mapper. So two promising paths after the PR was merged

File: core-bundle/contao/library/Contao/Image.php (hunk 7 above)
Comment (leofeyer) [reply core-bundle/contao/library/Contao/Image.php]: Restored in b9e5f051394692fedb1a90cf8e9dded6066e3d85.

File: core-bundle/src/Command/SymlinksCommand.php (hunk 9 above)
Comment (leofeyer) [reply core-bundle/src/Command/SymlinksCommand.php]: Changed in cf4ddeda99e78dcbaeb9859b7b6e9a54fdba62ab.

File: core-bundle/contao/library/Contao/Combiner.php [hunk 10]
```diff
@@ -212,9 +216,10 @@ public function getFileUrls($strUrl=null)
 
 				if ($blnDebug || !file_exists($this->strRootDir . '/' . $strPath))
```
Comment (ausi) [inline core-bundle/contao/library/Contao/Combiner.php]:
```suggestion
				if ($blnDebug || !file_exists($this->strRootDir . '/' . $this->strWebDir . '/' . $strPath))
```
As we dump into the web dir, the exists check probably needs to point there too?

Comment (ausi) [reply core-bundle/contao/library/Contao/Combiner.php]: Or maybe change the line above from `$strPath = 'assets/'…` to `$strPath = $this->strWebDir . '/assets/'…` and use `$this->strRootDir . '/' . $strPath` in the `dumpFile()` call below too.

**Review by ausi — COMMENTED** [review]
Do we need something that actively deletes an already existing symlink from `/public/assets` to `/assets`?
With a check like `!$this->symlinkAssets && is_link(Path::join($this->webDir, 'assets'))`?

File: core-bundle/contao/library/Contao/Combiner.php (hunk 10 above)
Comment (leofeyer) [reply core-bundle/contao/library/Contao/Combiner.php]: > As we dump into the web dir, the exists check probably needs to point there too?

Of course 🤦‍♂️

**Comment by leofeyer** [comment]
Added in 742e6473b5dc3ea394b220fc918a22eea8e34900.

File: core-bundle/src/Command/SymlinksCommand.php [hunk 11]
```diff
@@ -87,8 +96,15 @@ private function generateSymlinks(): void
         $this->symlinkModules();
         $this->symlinkThemes();
 
-        // Symlink the assets and themes directory
-        $this->symlink('assets', Path::join($this->webDir, 'assets'));
+        // Symlink the assets directory if the Contao components were installed there
+        // instead of in public/assets (backwards compatibility)
+        if ($this->symlinkAssets) {
+            $this->symlink('assets', Path::join($this->webDir, 'assets'));
+        } elseif (is_link($symlink = Path::join($this->webDir, 'assets'))) {
```
Comment (ausi) [inline core-bundle/src/Command/SymlinksCommand.php]:
```suggestion
        } elseif (is_link($symlink = Path::join($this->projectDir, $this->webDir, 'assets'))) {
```

Comment (leofeyer) [reply core-bundle/src/Command/SymlinksCommand.php]: Thank you

**Review by ausi — DISMISSED** [review]
LGTM 🎉
_(But I did not test it manually)_

**Comment by leofeyer** [comment]
I have tested it extensively and found another bug: ce48ebc5f69ee658f6ef357fc3c87fd2dbc36f79 Now everything should work.

**Review by ausi — APPROVED** [review]
